Lean on DB queries for describe-table for Mongo

calherries opened the pull request:
Resolves https://github.com/metabase/metabase/issues/46277

This PR reimplements `driver/describe-table` for MongoDB. Before we would query a sample of documents from a collection and analyse them in Clojure. Instead, we now now execute a query that does a similar aggregation, but most of the calculation is done in the Mongo database.

Based on a few tests the performance is slightly slower when the collection contains small or deeply nested documents but much faster for large ones. But the main difference is in memory usage. This uses very little memory in the Metabase instance because all of the aggregation is done in the database.

### Timing tests: 

Create the test collections in mongosh:
```
// oom_test

function generateLargeArrayDocument(arraySize) {
  const largeArray = Array.from({ length: arraySize }, (_, index) => ({
    id: index,
    value: Math.random()
  }));
  return {
    _id: new ObjectId(),
    largeArray: largeArray
  };
}

db.createCollection("oom_test")
const numDocuments = 500;
const arraySize = 100000;
for (let i = 0; i < numDocuments; i++) {
  db.oom_test.insertOne(generateLargeArrayDocument(arraySize));
}

// nested_data

function randomDate(start, end) {
    return new Date(start.getTime() + Math.random() * (end.getTime() - start.getTime()));
}

function generateDocument(id) {
    return {
        _id: id,
        string_field: `String value at depth 0 - ${id}`,
        number_field: Math.random() * 100,
        nested_object_1: {
            nested_1_string_field: `String value at depth 1 - ${id}`,
            nested_1_number_field: Math.random() * 100,
            nested_1_nested_object_1: {
                nested_1_nested_1_string_field: `String value at depth 2 - ${id}`,
                nested_1_nested_1_date_field: randomDate(new Date(2023, 0, 1), new Date(2025, 11, 31))
            },
            nested_1_nested_object_2: {
                nested_1_nested_2_string_field: `String value at depth 2 - ${id}`
            }
        },
        nested_object_2: {
            nested_2_string_field: `String value at depth 1 - ${id}`
        }
    };
}

db.nested_data.drop()
const numberOfDocuments = 1000;
let documents = [];
for (let i = 0; i < numberOfDocuments; i++) {
    documents.push(generateDocument(i));
}
db.nested_data.insertMany(documents);

// deep_data

function randomDate(start, end) {
    return new Date(start.getTime() + Math.random() * (end.getTime() - start.getTime()));
}

function generateDeepObject(depth, maxDepth) {
    if (depth >= maxDepth) {
        return {
            leaf_string: `string at depth ${depth}`,
            leaf_number: Math.random() * 100,
            leaf_date: randomDate(new Date(2023, 0, 1), new Date(2025, 11, 31))
        };
    }
    return {
        string_field: `string at depth ${depth}`,
        number_field: Math.random() * 100,
        nested_object: generateDeepObject(depth + 1, maxDepth)
    };
}

function generateDocument(id, maxDepth) {
    return {
        _id: id,
        top_level_string: "string",
        top_level_number: Math.random() * 100,
        deep_nested_object: generateDeepObject(0, maxDepth)
    };
}

db.deep_data.drop()
const numberOfDocuments = 1000;
const maxDepth = 175; // any higher exceeds the limit for nesting
let documents = [];
for (let i = 0; i < numberOfDocuments; i++) {
    documents.push(generateDocument(i, maxDepth));
}
db.deep_data.insertMany(documents);

// wide_deep_data
function generateWideDeepObject(width, depth, currentDepth = 0) {
    let obj = {};
    let objectWidth = [0, 10, 20, 30, 40].includes(currentDepth) ? width : 1;
    for (let i = 0; i < objectWidth; i++) {
        let key = `key_${currentDepth}_${i}`;
        if (currentDepth < depth) {
            obj[key] = generateWideDeepObject(width, depth, currentDepth + 1);
        } else {
            obj[key] = Math.random();
        }
    }
    return obj;
}

function generateWideDeepDocument(id, width, depth) {
    return {
        _id: id,
        data: generateWideDeepObject(width, depth)
    };
}

db.wide_deep_data.drop();
const numberOfDocuments = 1000;
const width = 4;  // Adjust this for wider documents
const depth = 50;  // Adjust this for deeper nesting
let documents = [];
for (let i = 0; i < numberOfDocuments; i++) {
    documents.push(generateWideDeepDocument(i, width, depth));
}
db.wide_deep_data.insertMany(documents);
```

Then sync the database and run the benchmarks in the repl:
```
(comment
  (def table (t2/select-one :model/Table :name "nested_data"))
  (def table (t2/select-one :model/Table :name "oom_test"))
  (def table (t2/select-one :model/Table :name "deep_data"))
  (def table (t2/select-one :model/Table :name "wide_deep_data"))
  (def db (t2/select-one :model/Database (:db_id table)))
  (criterium.core/quick-bench (do (driver/describe-table :mongo db table) nil)))
```
**Timing on my M1 laptop**:

nested_data (small and fairly typical shaped documents with some nesting)
- Before: 83 ms
- After: 154 ms

deep_data (deeply nested data, basically the worst-case scenario for the new method)
- Before: 4311 ms
- After: 1439 ms

oom_test (documents with large arrays)
- Before: 237572 ms (if the JVM doesn't have enough memory, this can OOM)
- After: 4116 ms

calherries committed:
~ use DB to describe the table

calherries committed:
~ tidy

calherries committed:
~ tidy

calherries committed:
~ optimize root query

calherries committed:
~ nested-level-query works and gets objects too

calherries committed:
+ root query gets objects too

calherries committed:
+ driver/describe-table :mongo works

calherries committed:
~ renames

calherries committed:
~ tidy

calherries committed:
~ tidy

calherries committed:
~ tidy

calherries committed:
~ fix index

calherries committed:
~ fix index

calherries committed:
~ sample first and last

calherries committed:
~ remove old implementation

calherries committed:
~ fix tests

calherries committed:
~ use set

calherries committed:
~ reduce memory footprint of sample

calherries committed:
~ fix

calherries committed:
~ k -> field

calherries committed:
~ use type/Instant

calherries committed:
~ fix test

calherries committed:
~ fix test

calherries committed:
~ fix root-query for nulls

calherries committed:
~ fix index

calherries committed:
~ fix test

calherries committed:
~ optimize

calherries committed:
~ fix

calherries committed:
~ tidy

calherries committed:
~ tidy

calherries committed:
~ tidy

calherries committed:
~ tidy

calherries committed:
~ tidy

calherries committed:
~ tidy

perivamsi commented:
Can we not backport this PR without having a discussion please?

calherries commented:
We're not going to backport this to 50 unless there is a customer that really needs it, and we get more confidence in it. https://metaboat.slack.com/archives/C0641E4PB9B/p1723648950115759?thread_ts=1723648266.831449&cid=C0641E4PB9B

camsaul commented on modules/drivers/mongo/test/metabase/driver/mongo_test.clj:
```diff
@@ -197,36 +196,44 @@
             (:tables (driver/describe-database :mongo (mt/db)))))))
 
-(deftest ^:parallel describe-table-test
+(defmacro with-describe-table-query-depths
```
```suggestion
(defmacro with-describe-table-query-depths!
```

To make it clear this is thread-unsafe and that it's not ok for use in parallel tests

camsaul commented on modules/drivers/mongo/src/metabase/driver/mongo.clj:
```diff
@@ -219,45 +145,215 @@
-      (log/error (format "Error introspecting collection: %s" (:name table)) t))))
+(defn- sample-stages
+  "Query stages which get a sample of the data in the collection, of size `n`. Half of the sample is from the first
+   inserted documents and the other half from the last inserted documents."
```
Nitpick https://guide.clojure.style/#docstring-indentation

Only pointing this out because enabling `cljfmt` against our codebase is something I'll probably work on in the near future as part of DevEx stuff and I don't want to have any more needless churn than we need

camsaul commented on modules/drivers/mongo/src/metabase/driver/mongo.clj:
```diff
@@ -219,45 +145,215 @@
-     (concat (sample-documents db table 1) (sample-documents db table -1)))
-    (catch Throwable t
-      (log/error (format "Error introspecting collection: %s" (:name table)) t))))
+(defn- sample-stages
```
A big comment block giving an overview of the code below and/or an example of the sort of query we build would really be nice. I really don't know what some of the stuff like `facet-stage` means TBH

camsaul commented on modules/drivers/mongo/test/metabase/driver/mongo_test.clj:
```diff
@@ -197,36 +196,44 @@
+     (with-redefs [mongo/describe-table-query-depth depth#]
+       ~@body)))
+
+(deftest describe-table-test
```
Would it be worth it to add a test just for `describe-table-query`, so given some certain input we can check what sort of native query we expect it to generate?

camsaul reviewed (approved):
Overall I'm 👍 on this change but I left some notes about maybe adding some more tests and comments and minor style stuff

calherries committed:
~ optimize query

calherries committed:
~ allow null

calherries committed:
+ make database-position deterministic

calherries committed:
~ tidy

calherries committed:
~ tidy

calherries committed:
~ fix renaming

calherries committed:
~ update test

calherries committed:
~ tidy

calherries committed:
Merge branch 'master' into mongo-db-describe-table

calherries committed:
~ fix

calherries committed:
~ fix

camsaul commented on modules/drivers/mongo/test/metabase/driver/mongo_test.clj:
```diff
@@ -197,32 +196,105 @@
             (:tables (driver/describe-database :mongo (mt/db)))))))
 
-(deftest ^:parallel describe-table-test
+(deftest describe-table-query-test
```
```suggestion
(deftest ^:parallel describe-table-query-test
```

camsaul reviewed (approved):
cool, this LGTM

calherries committed:
~ fix test

calherries committed:
~ make test parallel

calherries committed:
Merge branch 'master' into mongo-db-describe-table

calherries committed:
~ fix test

calherries committed:
Merge branch 'master' into mongo-db-describe-table

calherries commented:
Unfortunately this query is timing out on sharded clusters, so can't be merged yet. I’d recommend whoever is responsible for this side of things now takes a look at the PR and decides whether to throw it away, adjust it to work for sharded clusters, or use the old code just for sharded clusters and the new stuff for regular single-node DBs

snoe committed:
Various fixes for faster sync

Upgraded driver to 5.2.0
Updated data load to insert many rather than 1 row at a time.
Dropped max-depth to 7, see comment.

snoe committed:
Merge remote-tracking branch 'origin/master' into mongo-db-describe-table
